packages/vc-export/src/KiltAttestationProofV1.ts

```diff
@@ -0,0 +1,668 @@
+  const [events, time] = await apiAt.queryMulti<
+    [Vec<FrameSystemEventRecord>, u64]
+  >([
+    [api.query.system.events as QueryableStorageEntry<'promise'>],
+    [api.query.timestamp.now as QueryableStorageEntry<'promise'>],
+  ])
+
+  const timestamp = time.toNumber()
+  const attestationEvent = events
+    .reverse()
+    .find(
+      ({ phase, event }) =>
+        phase.isApplyExtrinsic &&
+        api.events.attestation.AttestationCreated.is(event) &&
+        u8aEq(event.data[1], claimHash)
+    )
+  if (!attestationEvent)
+    throw new SDKErrors.CredentialUnverifiableError(
+      `Matching attestation event for root hash ${u8aToHex(
+        claimHash
+      )} not found at block ${u8aToHex(blockHash)}`
+    )
+  const [att, , cTypeHash, authorization] = attestationEvent.event.data as [
+    AccountId,
+    Hash,
+    Hash,
+    Option<Hash> | Option<RuntimeCommonAuthorizationAuthorizationId>
+  ] &
+    IEventData
+  const attester = getFullDidUri(encodeAddress(att.toU8a(), 38))
+  const cTypeId = CType.hashToId(cTypeHash.toHex())
+  const delegationId = authorization.isSome
+    ? (
+        (authorization.unwrap() as RuntimeCommonAuthorizationAuthorizationId)
+          .value ?? authorization.unwrap()
+      ).toHex()
+    : null
+  return {
+    verified: true,
+    timestamp,
+    attester,
+    cTypeId,
+    delegationId,
+  }
+}
+
+async function findDelegators(
+  nodeId: string,
+  delegators: Set<DidUri>
+): Promise<void> {
+  const node = await DelegationNode.fetch(nodeId)
+  delegators.delete(node.account)
+  if (delegators.size === 0) {
+    return
+  }
+  if (node.parentId) {
+    await findDelegators(node.parentId, delegators)
+    return
+  }
+  throw new SDKErrors.CredentialUnverifiableError(
+    `The following delegators are not in the attestation's delegation hierarchy: ${[
+      ...delegators,
+    ]}`
+  )
+}
+
+/**
+ * Verifies a KILT attestation proof by querying data from the KILT blockchain.
+ * This includes querying the KILT blockchain with the credential id, which returns an attestation record if attested.
+ * This record is then compared against attester address and delegation id (the latter of which is taken directly from the credential).
+ *
+ * @param credentialInput Verifiable Credential to verify proof against.
+ * @param proof KiltAttestationProofV1 proof object to be verified. Any proofs embedded in the credentialInput are stripped and ignored.
+ * @param opts Additional parameters.
+ * @param opts.api A polkadot-js/api instance connected to the blockchain network on which the credential is anchored.
+ * @returns Object indicating whether proof could be verified.
+ */
+export async function verifyProof(
+  credentialInput: VerifiableCredential,
+  proof: KiltAttestationProofV1,
+  opts: { api?: ApiPromise } = {}
+): Promise<void> {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { proof: _, ...credential } = credentialInput
+  // 0. check proof structure
```

Comment:
Why don't we check the credential structure first? Is it more efficient to do so?

Reply:
Conceptually it makes sense to me; the proof needs to make assertions about its own structure; if it's not internally consistent, there is no point in checking whether it proves the integrity of the document for which it was created. And yes, it's probably also more efficient that way, as the proof is just a way smaller structure and thus cheaper to check.